pkg/db/v1beta1/postgres/postgres.go

```diff
@@ -0,0 +1,206 @@
+		return nil, fmt.Errorf("DB open failed: %v", err)
+	}
+	return NewWithSQLConn(db)
+}
+
+func NewWithSQLConn(db *sql.DB) (common.KatibDBInterface, error) {
+	d := new(dbConn)
+	d.db = db
+	seed, err := crand.Int(crand.Reader, big.NewInt(1<<63-1))
+	if err != nil {
+		return nil, fmt.Errorf("RNG initialization failed: %v", err)
+	}
+	// We can do the following instead, but it creates a locking issue
+	//d.rng = rand.New(rand.NewSource(seed.Int64()))
+	rand.Seed(seed.Int64())
+
+	return d, nil
+}
+
+func (d *dbConn) RegisterObservationLog(trialName string, observationLog *v1beta1.ObservationLog) error {
+	statement := "INSERT INTO observation_logs (trial_name, time, metric_name, value) VALUES "
+	values := []interface{}{}
+
+	index_of_qparam := 1
+	for _, mlog := range observationLog.MetricLogs {
+		if mlog.TimeStamp == "" {
+			continue
+		}
+		t, err := time.Parse(time.RFC3339Nano, mlog.TimeStamp)
+		if err != nil {
+			return fmt.Errorf("Error parsing start time %s: %v", mlog.TimeStamp, err)
+		}
+		sqlTimeStr := t.UTC().Format(time.RFC3339Nano)
+
+		statement += fmt.Sprintf("($%d, $%d, $%d, $%d),",
+			index_of_qparam, index_of_qparam+1, index_of_qparam+2, index_of_qparam+3,
+		)
+		values = append(values, trialName, sqlTimeStr, mlog.Metric.Name, mlog.Metric.Value)
+		index_of_qparam += 4
+	}
+
+	statement = statement[:len(statement)-1]
+
+	// Prepare the statement
+	stmt, err := d.db.Prepare(statement)
+	if err != nil {
+		return fmt.Errorf("Prepare SQL statement failed: %v", err)
+	}
+
+	// Defer Close the statement
+	defer stmt.Close()
+
+	// Execute INSERT
+	_, err = stmt.Exec(values...)
+	if err != nil {
+		return fmt.Errorf("Execute SQL INSERT failed: %v", err)
+	}
+
+	return nil
+}
+
+func (d *dbConn) GetObservationLog(trialName string, metricName string, startTime string, endTime string) (*v1beta1.ObservationLog, error) {
+	qfield := []interface{}{trialName}
+	qstr := ""
+	index_of_qparam := 1
+
+	base_stmt := fmt.Sprintf("SELECT time, metric_name, value FROM observation_logs WHERE trial_name = $%d", index_of_qparam)
+	index_of_qparam += 1
+
+	if metricName != "" {
+		qstr += fmt.Sprintf(" AND metric_name = $%d", index_of_qparam)
+		qfield = append(qfield, metricName)
+		index_of_qparam += 1
+	}
+
+	if startTime != "" {
+		s_time, err := time.Parse(time.RFC3339Nano, startTime)
+		if err != nil {
+			return nil, fmt.Errorf("Error parsing start time %s: %v", startTime, err)
+		}
+		formattedStartTime := s_time.UTC().Format(time.RFC3339Nano)
+		qstr += fmt.Sprintf(" AND time >= $%d", index_of_qparam)
+		qfield = append(qfield, formattedStartTime)
+		index_of_qparam += 1
+	}
+	if endTime != "" {
+		e_time, err := time.Parse(time.RFC3339Nano, endTime)
+		if err != nil {
+			return nil, fmt.Errorf("Error parsing completion time %s: %v", endTime, err)
+		}
+		formattedEndTime := e_time.UTC().Format(time.RFC3339Nano)
+		qstr += fmt.Sprintf(" AND time <= $%d", index_of_qparam)
+		qfield = append(qfield, formattedEndTime)
+		// i += 1
```

Comment:
Just a question.
Why is it not necessary to perform `index_of_qparam += 1`?

Reply:
I have solved myself.

BTW, should we change this comment?

```suggestion
		// index_of_qparam += 1
```

Reply:
As you might solved yourself, it is for future (if any other query filter added). I'll add comments about it. Thanks for raising this!

Reply:
Thanks for clarifying! @anencore94